File: threeSumClosest.cpp
#include<iostream>
#include<string>
#include<vector>
#include<algorithm>
using namespace std;

//#define test

#ifdef test
void testFunc()
{
    int t=-1;
    cout<<abs(t)<<endl;
}
#endif

bool cmp(int a,int b)
{
    return a < b; //升序
}

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(begin(nums),end(nums),cmp);
        int high,low;
        int tmp;
        int sum=1<<25;
        
        for(unsigned long long i=0;i<nums.size()-2;i++)
        {
            low=i+1;
            high=nums.size()-1;
            while(low<high)
            {
                tmp=nums[high]+nums[low]+nums[i];
                if(abs(tmp-target) < abs(sum-target))
                    sum=tmp;
                if(tmp < target)
                    low++;
                else
                    high--;
            }
        }
        return sum;
    }
};

int main()
{
#ifdef test
    testFunc();
#endif
    Solution sl;
    vector<int> in = {-1,2,1,-4};
    int out = sl.threeSumClosest(in,1);
    cout<<out<<endl;
    return 0;
}

File: threeSumClosest.py
from typing import List


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        res = 0
        d = 1234567
        nums.sort()

        for i in range(len(nums) - 2):
            k = len(nums) - 1

            for j in range(i + 1, len(nums) - 1):
                t = nums[i] + nums[j] + nums[k]
                nd = abs(t - target)

                while t >= target:
                    if nd < d:
                        print(i, j, k, t, nd, d)
                        d = nd
                        res = t
                    k -= 1

                    if k <= j:
                        break
                    t = nums[i] + nums[j] + nums[k]
                    nd = abs(t - target)

                if nd < d:
                    # print(i,j,k,t,nd,d)
                    d = nd
                    res = t

                if k <= j:
                    break

        return res


nums = [-1, 2, 1, -4]
target = 1
nums = [0, 0, 0]
target = 1
print(Solution().threeSumClosest(nums, target))
